feat(new_module): Static and moving vision-based target esitmator (Kalman Filter) (#23726)
PX4/PX4-Autopilot: docs/en/advanced_features/vision_target_estimator.md

# Vision Target Estimator (VTEST)

> The Vision Target Estimator (VTEST) fuses vision-based relative target measurements with absolute positioning sources (vehicle GNSS, a mission landing waypoint, and/or target-mounted GNSS).

Reviewer: This is a very abrupt introduction. I think you're expecting your reader to be an expert engineer in this field, possibly with PhD. The target for this should be an educated person. At the top you can't assume a lot of knowledge of PX4, but futher down you can. So "It runs alongside the primary vehicle estimator and publishes:" is too much at this level.

 The very first thing you should do is identify the audience by saying what this is and who it is for.

I'm guessing from the name it is something that allows you to esimate the position/post of some target in front of you based on some kind of visual hints? What kind of hints? When would you use this - for precision landing? 

Then identify audience "this is for developers who want to Xxxx"

If this is going to be at this depth up front you need to push it somewhere no one who is starting is going to see it. That's why something like https://docs.px4.io/main/en/advanced_config/tuning_the_ecl_ekf is in the depths. It's a reference once you've done all the other tuning possible.

Reviewer: PS I have subedited for style, typos etc and added a few comments. I haven't actually read it though. I will try do so next week when I understand what this is for.

Author: Agreed, updated the intro and explained whe this page is meant for: 

The Vision Target Estimator (VTEST) estimates where a target is, relative to the vehicle, by combining a camera-based detection of the target (for example an ArUco fiducial marker) with one or more absolute position references: the vehicle's own GNSS, the mission landing waypoint, and/or a GNSS receiver mounted on the target itself.
Its main use is [precision landing](../advanced_features/precland.md), where the vehicle needs to touch down on a small, well-defined spot rather than the approximate mission waypoint.

This page is for developers who want to enable VTEST on a custom build, tune it for their vehicle, or integrate a new vision pipeline that publishes landing-target observations.
It assumes familiarity with PX4 and basic state estimation, but does not require prior knowledge of the module itself.
For tuning workflows, log analysis guidance, and instructions on extending the estimator, see the [Vision Target Estimator deep dive](../advanced_features/vision_target_estimator_advanced.md).

::: warning
VTEST is a beta feature, disabled in default board configurations, and should only be enabled on custom builds after careful bench and flight testing.
:::

Reviewer: That is perfect ^^^. I haven't given the whole thing a review, but that is enough context to provide relevance.

It runs alongside the primary vehicle estimator and publishes:

- `landing_target_pose` - relative and absolute target pose in NED for precision landing.
- `vision_target_est_position` - the full position-state vector (with covariances) for logging and tuning.
- `vision_target_est_orientation` - target yaw and yaw-rate estimates with associated uncertainty.
- `vision_target_est_input` - downsampled NED acceleration and attitude quaternion used by the position prediction step.
- `vte_bias_init_status` - raw and low-pass filtered GNSS/vision bias, plus the consecutive raw-bias delta norm, during the initial bias averaging phase.
- `vte_aid_*` - innovation logs per sensor, used for observability checks and noise tuning.

When [`VTE_EKF_AID`](../advanced_config/parameter_reference.md#VTE_EKF_AID)=1 and the target is static, the relative velocity estimate is forwarded to the main EKF2 estimator (see [EKF2 aiding](#ekf2-aiding) for the required conditions).

> > [!WARNING]
> > VTEST is a beta feature, disabled in default board configurations, and should only be enabled on custom builds after careful bench and flight testing.

Reviewer: We don't use this markdown flavour. It works, but it doesn't localise properly.

```suggestion
::: warning
VTEST is a beta feature, disabled in default board configurations, and should only be enabled on custom builds after careful bench and flight testing.
:::
```

Author: updated, also for > [!TIP] --> ::: tip

For tuning workflows, log analysis guidance, and instructions on extending the estimator, see the [Vision Target Estimator deep dive](../advanced_features/vision_target_estimator_advanced.md).

## Table of Contents

> [[toc]]

Reviewer: I'm not a huge fan of in-page TOCs, but since this is such a big doc I guess it makes sense. I have replaced this with the macro though so you don't have to maintain it. Same in other doc.

Author: Thanks

# Vision Target Estimator (VTEST)

> 

Reviewer: Actually, let's badge this

```suggestion

<Badge type="tip" text="PX4 v1.18" /> <Badge type="warning" text="Experimental" />

```

> # Vision Target Estimator (VTEST)
> 
> The Vision Target Estimator (VTEST) estimates where a target is, relative to the vehicle, by combining a camera-based detection of the target (for example an ArUco fiducial marker) with one or more absolute position references: the vehicle's own GNSS, the mission landing waypoint, and/or a GNSS receiver mounted on the target itself.

Reviewer: ```suggestion
# Vision Target Estimator (VTEST)

<Badge type="tip" text="PX4 v1.18" /> <Badge type="warning" text="Experimental" />

The Vision Target Estimator (VTEST) estimates where a target is, relative to the vehicle, by combining a camera-based detection of the target (for example an ArUco fiducial marker) with one or more absolute position references: the vehicle's own GNSS, the mission landing waypoint, and/or a GNSS receiver mounted on the target itself.
```

> # Vision Target Estimator (VTEST)

Reviewer: @JonasPerolini I went through this at high level again. Things I did:
- Added a few comments (and will add some more)
- Moved this stuff here out of the introduction. More on that to follow.
- Tidied up the "Building the Module" a bit. I wanted to make clear in simple list what the modules/settings are. In particular that we'll need to come back and fix up the key for mavlink.
  That's an important note. When this is close to acceptance you should go to MAVLink and create a PR to move the messages from development.xml to common.xml, with a clear explanation of what you are doing them and how tested they are - i.e. if you don't use some feature of them then it is untested, and we should be clear on that.
- Tidied up the MAVLink intro for same reason.
- Ran prettier
- Cross linked to the uorb topics that WILL exist when you merge this from the first instance they appear in the docs (makes it easier for readers to know exactly what you are talking about). 
  You can build these if you like because no one else can swamp them.
- Regenerated the parameter_reference.md. This still needs to be deleted before merging, but it allows us to test links now.

Most readers are impatient.
They want to know the minimum to get something up and running.
They might be interested in technical details, but only as far as they need to know them to configure the software.
For that reason it is usually best to have a very focussed document on setup and configuration, designed to be as turnkey as possible.
This might be followed in the same doc with a deeper section on configuration with more explanation, or even another document.
As good exmaple of this, consider the work we've been doing on the ADSB/Avoidance library you are working on. THe user doens't need to know much to configure it broadly - so all the detail is in a separate doc.

What I'd like you to do is see if we can do something similar here.
Either push more of the content into the deep dive, or push the detail into a separate section that is still relevant at this level.
THIS might not be possible, but I'd like you to try!

Questions to ask yourself:
- Does the reader need to know this to configure the system - do they NEED an architectural overview. What about the operational overview. If they don't NEED it, is it still USEFUL for them?
  IF the answer is not, and maybe just a little, then we might push it into the deep dive.
- When explaining how to set a configuration option, tell the user what they need to do, not what just the option does. So, if you have a configuration option with a default value and most users won't need to change it, say that. If they might need to change it, explain how to do so, not what the values are. I'll give you an example in place.

## How would I do this?

For the basic doc, I think you just need
- Overview/intro - as is
- Building the Module
- Configuration
- Gazebo Classic Simulation
- Monitoring
- Operational Notes

I could be wrong, but ideally these would disappear to the end or into the advanced guide.
- Operational Overview
- Architecture Overview
- Dynamic Models
- Estimator Lifecycle
- Measurement Sources

Does this make sense?

Author: Thanks for the review @hamishwillee, updated here: https://github.com/PX4/PX4-Autopilot/pull/23726/changes/53fb99d66f5e9a272a6247198dcac814061db869

Basic docs:
- Deleted Operational Overview: the uORB output list duplicated content already in the advanced doc's Estimator Outputs.
- Merged Architecture Overview + Dynamic Models + Estimator Lifecycle into a new Estimator Overview section:
   - A few paragraphs explaining: two filters (position + orientation) at 50 Hz, three 1D KFs per NED axis, what a prediction step and an update step do, when fusion starts, task activation, and a pointer to the timeout config.
   - Sub-paragraphs: Dynamic Models (static state only, with a link to the moving-target section in advanced), Bias Estimation, Time Alignment. 
   - Removed internal file/class names (VisionTargetEst, VTEPosition, kPosUpdatePeriodUs, etc.).
- Outlier detection shortened into a single sentence inside the intro instead of its own subsection.
- Building the Module shortened: removed the moving-target build keys and targets (now in advanced), keeping only the static path.
- Deleted Measurement Validity subsection (its parameters are already listed in Module Enable).
- Deleted Moving-target projection parameters (experimental) subsection entirely (moved to advanced).

Advanced docs: 
- Added a new Moving-target Mode (experimental) section with everything moved out of basics docs: the 5-state dynamic model with target velocity/acceleration, the build setup (CONFIG_VTEST_MOVING=y + the two prebuilt make targets),the precision-landing parameters, ...

Author: This commit: https://github.com/PX4/PX4-Autopilot/pull/23726/changes/be976d9115b15e26b970a782638e88e014e50ef4 provides more info to the configuration part, why and when to change the defaults.

  This bias is the offset between the active absolute target reference (target-mounted GNSS or the mission landing waypoint) and the vision-based target position, so the absolute reference can still point to the target if vision is lost.
  When the firmware is built with `CONFIG_VTEST_MOVING`, target velocity and acceleration states are also estimated (experimental).
- **Orientation filter (`VTEOrientation`)** - a planar yaw filter that estimates the target's heading and yaw rate.

Both filters run at a fixed 20 ms interval (`kPosUpdatePeriodUs = kYawUpdatePeriodUs`), so the target rate is 50 Hz.
The position loop downsamples vehicle acceleration, rotates it into NED, forwards attitude/local position/local velocity data to `VTEPosition`, and publishes `vte_input`.
The yaw loop advances `VTEOrientation` on the same scheduler cadence using its own measurement subscriptions.
The work item also enforces timeouts, parameter reloads, and task activation.

::: warning
The moving-target mode `CONFIG_VTEST_MOVING=y` is experimental and has not yet been flight tested.

> :::

Reviewer: What does a developer setting up their system need to know of this?

Author: Simplified in https://github.com/PX4/PX4-Autopilot/pull/23726/changes/53fb99d66f5e9a272a6247198dcac814061db869, it is now basic and high level, no more file names or code constants

The keys and values that need to be present are:

- `CONFIG_MODULES_VISION_TARGET_ESTIMATOR=y`: _Enable_ VTEST in firmware
- `CONFIG_MODULES_LANDING_TARGET_ESTIMATOR=n`: _Disable_ the landing target estimator (both modules publish [`landing_target_pose`](../msg_docs/LandingTargetPose.md) and will conflict if enabled together).
- `CONFIG_VTEST_MOVING=y`: (Optionally) Enable experimental support for estimation of target velocity and acceleration states.
- `CONFIG_MAVLINK_DIALECT="development"`: Enable using the development dialect.
  Note that this step will no longer be required once the MAVLink messages used by VTEST have been validated.

The following build targets define custom configurations that include the requied changes:

- `make px4_fmu-v6c_visionTargetEstStatic`: static-target build (`CONFIG_VTEST_MOVING=n`).
- `make px4_fmu-v6c_visionTargetEstMoving`: experimental moving-target build (`CONFIG_VTEST_MOVING=y`).

You can use these targets to test on SITL:

- `make px4_sitl_default`: SITL static-target build.
- `make px4_sitl_vtest-moving`: SITL experimental moving-target build (`CONFIG_VTEST_MOVING=y`).

> ## Dynamic Models

Reviewer: Does any of this affect how a user would configure the system? 

To me this feels like the shallow bit of the deep dive.

Author: Yes it is usefull, I've moved the moving filter to the advanced docs, it is now easier to read with only the static equations

   When GNSS is active first, VTEST low-pass filters the early raw samples (tuned by [`VTE_BIA_AVG_THR`](../advanced_config/parameter_reference.md#VTE_BIA_AVG_THR) and [`VTE_BIA_AVG_TOUT`](../advanced_config/parameter_reference.md#VTE_BIA_AVG_TOUT)) so that vision is only fused once the offset has settled.
   See [Bias initialization design](../advanced_features/vision_target_estimator_advanced.md#bias-initialization-design) for the full state-reset rules, the LPF exit condition, and the stale-GNSS fallback.
4. **Prediction**: Predictions run whenever the fixed 20 ms scheduler interval elapses (50 Hz).
5. **Update**: When new sensor observations are available, measurement handlers check the validity of the measurement (timestamp, value, and fusion requested in [`VTE_AID_MASK`](../advanced_config/parameter_reference.md#VTE_AID_MASK)), process it, and populate observation buffers.
   The observations that pass the NIS gates defined by [`VTE_POS_NIS_THRE`](../advanced_config/parameter_reference.md#VTE_POS_NIS_THRE) and [`VTE_YAW_NIS_THRE`](../advanced_config/parameter_reference.md#VTE_YAW_NIS_THRE) are then fused.

> 6. **Timeout handling**: If no measurement is fused for [`VTE_BTOUT`](../advanced_config/parameter_reference.md#VTE_BTOUT), the affected filter is reset and can restart immediately once its task and enabled fusion sources are available; the published pose/yaw is flagged invalid once [`VTE_TGT_TOUT`](../advanced_config/parameter_reference.md#VTE_TGT_TOUT) elapses without fresh data.

Reviewer: Interesting, but another bit of the deep dive IMO. The decision wether this goes top level or lower is around how much most readers need to know to use the feature.

Author: This was simplified here https://github.com/PX4/PX4-Autopilot/pull/23726/changes/53fb99d66f5e9a272a6247198dcac814061db869, keeping only the most relevant and high level info

::: info
UWB and IRLock are candidates for future development once representative test data is available.
:::

### Measurement Validity

Two time horizons guard incoming data: [`VTE_M_REC_TOUT`](../advanced_config/parameter_reference.md#VTE_M_REC_TOUT) defines how old a sample can be and still be considered for fusion, while [`VTE_M_UPD_TOUT`](../advanced_config/parameter_reference.md#VTE_M_UPD_TOUT) bounds how long cached measurements remain valid in the estimator state.
The published outputs toggle `*_valid` false after [`VTE_TGT_TOUT`](../advanced_config/parameter_reference.md#VTE_TGT_TOUT), and the affected estimator resets and retries immediately once [`VTE_BTOUT`](../advanced_config/parameter_reference.md#VTE_BTOUT) is exceeded.
Measurements timestamped in the future relative to the latest prediction are always rejected and reported on the innovation topics.

> ## Configuration

Reviewer: This is the important bit in this doc.

This is where we decide how much of the background you need, because sometimes you can't configure without an understanding of the architecture.

Author: Added more info here: https://github.com/PX4/PX4-Autopilot/pull/23726/changes/be976d9115b15e26b970a782638e88e014e50ef4

### Module Enable and Scheduling

- [`VTE_EN`](../advanced_config/parameter_reference.md#VTE_EN) - global module enable (reboot required).
- [`VTE_POS_EN`](../advanced_config/parameter_reference.md#VTE_POS_EN) / [`VTE_YAW_EN`](../advanced_config/parameter_reference.md#VTE_YAW_EN) - enable the position and orientation filters respectively (reboot required).
- [`VTE_BTOUT`](../advanced_config/parameter_reference.md#VTE_BTOUT), [`VTE_TGT_TOUT`](../advanced_config/parameter_reference.md#VTE_TGT_TOUT) - timeouts for estimator reset/retry and published validity flags.

> - [`VTE_M_REC_TOUT`](../advanced_config/parameter_reference.md#VTE_M_REC_TOUT), [`VTE_M_UPD_TOUT`](../advanced_config/parameter_reference.md#VTE_M_UPD_TOUT) - maximum ages for measurements to be fused or retained.

Reviewer: I think I'd separate the scheduling from the module enable, and move the "Measurement Validity" section from above under here to explain how these two values should be set. 

So not so much this is what they are "maximum ages for measurements to be fused or retained.". But "What value shoudl I tell the user to set for these"? What are the considerations? Can I just use the default.

Author: Agreed, split into 
- Module Enable
- Timeouts
- Measurement Freshness

here: https://github.com/PX4/PX4-Autopilot/pull/23726/changes/be976d9115b15e26b970a782638e88e014e50ef4 and explained when and why the params should be changed

:::

### Sensor Fusion Selection

[`VTE_AID_MASK`](../advanced_config/parameter_reference.md#VTE_AID_MASK) defines which measurements can be fused:

| Bit | Value | Meaning                                 |
| --- | ----- | --------------------------------------- |
| 0   | `1`   | Target GNSS position                    |
| 1   | `2`   | Vehicle GNSS velocity                   |
| 2   | `4`   | Vision-relative position                |
| 3   | `8`   | Mission landing waypoint                |
| 4   | `16`  | Target GNSS velocity (moving mode only) |

Bit 2 also enables processing of `fiducial_marker_yaw_report` in the orientation filter.

### Noise and Gating

- Adjust [`VTE_POS_UNC_IN`](../advanced_config/parameter_reference.md#VTE_POS_UNC_IN), [`VTE_VEL_UNC_IN`](../advanced_config/parameter_reference.md#VTE_VEL_UNC_IN), [`VTE_BIA_UNC_IN`](../advanced_config/parameter_reference.md#VTE_BIA_UNC_IN), [`VTE_ACC_UNC_IN`](../advanced_config/parameter_reference.md#VTE_ACC_UNC_IN), and [`VTE_YAW_UNC_IN`](../advanced_config/parameter_reference.md#VTE_YAW_UNC_IN) to reflect initial uncertainty.

>   These parameters seed the estimator state at initialization.

Reviewer: In this kind of sectio you need to capture
- How should I adjust them? 
- How do I know that I need to adjust them, or would benefit from adusting them?
- What are the considerations ?

You have some of the considerations below "Larger numbers slow initial convergence but reduce the chance of aggressive transients" which is great. Would be good to give people an idea of the approach such as "start with the default value, you know it is a problem if you see  [XXXX] or [YYYY]. In that case lower the value to XXX, kind of thing.

Author: Added more info on how to tune here https://github.com/PX4/PX4-Autopilot/pull/23726/changes/be976d9115b15e26b970a782638e88e014e50ef4

`mavlink_receiver` validates the frame/type and handles the message differently depending on the Vision Target Estimator status:

- When `VTE_EN=0`, the measurement is rotated (using `q_sensor` or the vehicle attitude for body-frame reports) and published straight to `landing_target_pose` so precision-landing can operate without the estimator.
- When `VTE_EN=1`, the message is split into `fiducial_marker_pos_report` and `fiducial_marker_yaw_report`.
  `VisionTargetEst` consumes these uORB topics to drive the position and orientation filters (using `fiducial_marker_pos_report.q` to rotate `fiducial_marker_pos_report.rel_pos` into NED at `timestamp_sample`).

#### TARGET_ABSOLUTE (ID 510)

`TARGET_ABSOLUTE` reports the target's absolute state when it carries its own GNSS (and optionally IMU).
A capability bitmap advertises which fields are valid.
PX4 maps the available content into the `target_gnss` uORB topic:

- Bit 0 (position) triggers publication of latitude/longitude/altitude along with the horizontal and vertical accuracy estimates (`position_std`).
- Bit 1 (velocity) forwards the NED velocity vector (`vel`) and its standard deviations (`vel_std`).

> - Additional fields (acceleration, quaternion `q_target`, rates, uncertainties) are not supported and reserved for future fusion logic once flight testing is available.

Reviewer: These are interesting, in that any setup will need to set them and send them. In particular this is good because it states what bits are used and how they are used.

You might consider updating this to make it clear what data rates PX4 expects for these messges (if any), whether the uOrb topics could also be used to feed the data if someone wanted to use ROS 2, any expections on the companion or camera that is providing it.

Author: added data rate info + uorb topics here https://github.com/PX4/PX4-Autopilot/pull/23726/changes/be976d9115b15e26b970a782638e88e014e50ef4

### Moving-target projection parameters (experimental)

When `CONFIG_VTEST_MOVING` is active, the estimator publishes the current target pose and velocity only.
The precision-landing controller applies the moving-target lookahead using [`PLD_MOVING_T_MIN`](../advanced_config/parameter_reference.md#PLD_MOVING_T_MIN) and [`PLD_MOVING_T_MAX`](../advanced_config/parameter_reference.md#PLD_MOVING_T_MAX).
It computes an intersection time $\Delta t$ by constraining $|z_{rel}| / |v_{descent}|$ between the two parameters, then shifts the horizontal landing setpoint by $\Delta t$ along the estimated target velocity.
The goal is to command the vehicle towards where the moving target will be at touchdown, not where it was at the last measurement.

> ## Gazebo Classic Simulation

Reviewer: Cool. Do you know if Gazebo will have support? Gazebo classic will be dead soon for PX4 IMO.

Author: I added the support for vision-based detection here: https://github.com/PX4/PX4-SITL_gazebo-classic/pull/950 so we will probably need to port it to the new version
